File: src/core/catalogue.ts
import googleJson from '../data/google-fonts.json'
import fontsourceJson from '../data/fontsource-fonts.json'
import { normalizeKey, parseFontName } from './names'
import type { Category, Family, FamilyFile, Source } from './types'

/**
 * The unified catalogue: Google Fonts plus the Fontsource families Google
 * does not carry, in one shape, with search and filtering.
 *
 * Both catalogues are build-time snapshots (see scripts/). Downloads happen
 * at runtime from CORS-enabled hosts:
 *
 *   Google      raw.githubusercontent.com/google/fonts — the original, complete,
 *               unsubsetted .ttf files, exactly what is installable. Mirrored by
 *               cdn.jsdelivr.net/gh, which is tried when GitHub rate-limits.
 *   Fontsource  cdn.jsdelivr.net/fontsource — single-subset families only, for
 *               which the one file is the entire font. See the build script for
 *               why a multi-subset Fontsource TTF must never be shipped.
 *
 * NOT the Google CSS API for downloads: it serves subsetted woff2 to a browser
 * UA and a subsetted font installs cleanly and then renders blanks. The CSS
 * API is used for previews only, in `src/platform/preview.ts`.
 */

interface RawGoogle {
  n: string
  c: string
  w: number[]
  i: number
  l: string
  p: string
  f: string[]
  b: number
  s: string[]
  a: string[]
  d: string[]
  r: number
  t: string
}

interface RawFontsource {
  n: string
  s: string
  u: string
  c: string
  w: number[]
  i: number
  l: string
  v: number
  e: string
  t: string
}

const RAW_BASE = 'https://raw.githubusercontent.com/google/fonts/main'
const GH_MIRROR = 'https://cdn.jsdelivr.net/gh/google/fonts@main'
const FONTSOURCE_CDN = 'https://cdn.jsdelivr.net/fontsource/fonts'

/** Licence text filename by repo directory. 38 of 44 apache/ families carry one. */
const LICENSE_FILE: Record<string, string> = { ofl: 'OFL.txt', apache: 'LICENSE.txt', ufl: 'UFL.txt' }

export const CATALOGUE_DATES = {
  google: (googleJson as { generated: string }).generated,
  fontsource: (fontsourceJson as { generated: string }).generated,
}

/** Style token as it appears in a static filename, e.g. `Poppins-SemiBold.ttf`. */
const STYLE_WEIGHTS: Array<[string, number]> = [
  ['Thin', 100],
  ['Hairline', 100],
  ['ExtraLight', 200],
  ['UltraLight', 200],
  ['Light', 300],
  ['Regular', 400],
  ['Book', 400],
  ['Medium', 500],
  ['SemiBold', 600],
  ['DemiBold', 600],
  ['ExtraBold', 800],
  ['UltraBold', 800],
  ['Bold', 700],
  ['Black', 900],
  ['Heavy', 900],
]

/** Weight and style of a google/fonts static filename. */
export function describeGoogleFile(file: string): { weight: number; italic: boolean; variable: boolean } {
  const base = file.replace(/\.ttf$/i, '')
  const variable = /\[[^\]]+\]$/.test(base)
  const stem = base.replace(/\[[^\]]+\]$/, '')
  const suffix = stem.includes('-') ? stem.slice(stem.indexOf('-') + 1) : ''
  const italic = /italic/i.test(suffix)
  const withoutItalic = suffix.replace(/italic/i, '')
  let weight = 400
  for (const [token, w] of STYLE_WEIGHTS) {
    if (withoutItalic.toLowerCase() === token.toLowerCase()) {
      weight = w
      break
    }
  }
  return { weight, italic, variable }
}

function category(raw: string): Category {
  const k = raw.toLowerCase().replace(/[^a-z]/g, '')
  if (k === 'sansserif') return 'Sans Serif'
  if (k === 'serif') return 'Serif'
  if (k === 'display') return 'Display'
  if (k === 'handwriting') return 'Handwriting'
  if (k === 'monospace') return 'Monospace'
  return 'Other'
}

function slug(name: string): string {
  return name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '')
}

function fromGoogle(f: RawGoogle): Family {
  const files: FamilyFile[] = f.f.map((file) => {
    const d = describeGoogleFile(file)
    const enc = encodeURIComponent(file)
    return {
      filename: file,
      url: `${RAW_BASE}/${f.p}/${enc}`,
      mirrors: [`${GH_MIRROR}/${f.p}/${enc}`],
      ...d,
    }
  })
  return {
    id: `google:${slug(f.n)}`,
    name: f.n,
    source: 'google',
    category: category(f.c),
    weights: f.w,
    italic: f.i === 1,
    license: f.l || 'Unknown',
    licenseUrl: f.p ? `${RAW_BASE}/${f.p}/${LICENSE_FILE[f.p.split('/')[0]!] ?? 'OFL.txt'}` : null,
    subsets: f.s,
    axes: f.a,
    designers: f.d,
    rank: f.r,
    added: f.t,
    bytes: f.b,
    files,
    key: normalizeKey(f.n),
  }
}

/**
 * Fontsource filenames on the CDN are `{subset}-{weight}-{style}.ttf`, which
 * says nothing about which font it is. They are renamed to the usual
 * `Family-Weight.ttf` convention, so a folder of them is usable.
 */
const WEIGHT_NAMES: Record<number, string> = {
  100: 'Thin',
  200: 'ExtraLight',
  300: 'Light',
  400: 'Regular',
  500: 'Medium',
  600: 'SemiBold',
  700: 'Bold',
  800: 'ExtraBold',
  900: 'Black',
}

export function fontsourceFilename(family: string, weight: number, italic: boolean): string {
  const stem = family.replace(/[^A-Za-z0-9]+/g, '')
  const w = WEIGHT_NAMES[weight] ?? String(weight)
  const style = italic ? (weight === 400 ? 'Italic' : `${w}Italic`) : w
  return `${stem}-${style}.ttf`
}

function fromFontsource(f: RawFontsource): Family {
  const files: FamilyFile[] = []
  const styles = f.i ? [false, true] : [false]
  const version = f.e && f.e !== 'latest' ? f.e : 'latest'
  // Fontsource has no size in its list endpoint; ~120 KB a face is typical.
  for (const w of f.w.length ? f.w : [400]) {
    for (const italic of styles) {
      files.push({
        filename: fontsourceFilename(f.n, w, italic),
        url: `${FONTSOURCE_CDN}/${f.s}@${version}/${f.u}-${w}-${italic ? 'italic' : 'normal'}.ttf`,
        mirrors: [],
        weight: w,
        italic,
        variable: false,
      })
    }
  }
  return {
    id: `fontsource:${f.s}`,
    name: f.n,
    source: 'fontsource',
    category: category(f.c),
    weights: f.w,
    italic: f.i === 1,
    license: f.l,
    // Every Fontsource package ships its LICENSE file on npm.
    licenseUrl: `https://cdn.jsdelivr.net/npm/@fontsource/${f.s}@${version}/LICENSE`,
    subsets: [f.u],
    axes: [],
    designers: [],
    // Google ranks its own; anything from Fontsource sorts after all of them,
    // alphabetically, rather than pretending to a popularity it was never given.
    rank: 100000,
    added: f.t,
    bytes: files.length * 120_000,
    files,
    key: normalizeKey(f.n),
  }
}

const GOOGLE = (googleJson as { families: RawGoogle[] }).families.map(fromGoogle)
const FONTSOURCE = (fontsourceJson as { families: RawFontsource[] }).families.map(fromFontsource)

/** Every family, Google first then Fontsource, each block alphabetical. */
export const FAMILIES: readonly Family[] = [...GOOGLE, ...FONTSOURCE]

export const COUNTS = {
  google: GOOGLE.length,
  googleDownloadable: GOOGLE.filter((f) => f.files.length > 0).length,
  fontsource: FONTSOURCE.length,
  total: FAMILIES.length,
  files: FAMILIES.reduce((n, f) => n + f.files.length, 0),
  bytes: FAMILIES.reduce((n, f) => n + f.bytes, 0),
}

const BY_ID = new Map<string, Family>()
const BY_KEY = new Map<string, Family>()
for (const f of FAMILIES) {
  BY_ID.set(f.id, f)
  // Google wins a name collision, by construction of the Fontsource catalogue
  // (it excludes anything Google has) — but be explicit, not lucky.
  if (!BY_KEY.has(f.key)) BY_KEY.set(f.key, f)
}

export function familyById(id: string): Family | undefined {
  return BY_ID.get(id)
}

/**
 * Exact family lookup by name, tolerant of case, punctuation and a trailing
 * style word: `poppins`, `Poppins-Bold` and `Poppins SemiBold Italic` all
 * find Poppins. A name that is *only* a style word (`Black`) is kept whole.
 */
export function findFamily(name: string, source: Source | 'auto' = 'auto'): Family | undefined {
  const candidates = [normalizeKey(name), normalizeKey(parseFontName(name).family)]
  for (const key of candidates) {
    const hit = BY_KEY.get(key)
    if (!hit) continue
    if (source === 'auto' || hit.source === source) return hit
  }
  if (source !== 'auto') {
    // A specific source was asked for and the key-map's winner is the other
    // one. Scan that source only.
    const pool = source === 'google' ? GOOGLE : FONTSOURCE
    for (const key of candidates) {
      const hit = pool.find((f) => f.key === key)
      if (hit) return hit
    }
  }
  return undefined
}

/**
 * Near matches for a name that found nothing, for the list importer.
 *
 * Token overlap rather than edit distance: the useful suggestions are
 * "Garamond" -> "EB Garamond" / "Cormorant Garamond", which share a whole
 * word but are far apart by character edits.
 */
export function suggestFamilies(name: string, limit = 3): Family[] {
  const wanted = normalizeKey(parseFontName(name).family)
  const wantedTokens = new Set(wanted.split(' ').filter((t) => t.length > 2))
  if (wantedTokens.size === 0) return []

  const scored: Array<{ f: Family; score: number }> = []
  for (const f of FAMILIES) {
    if (f.key === wanted) continue
    const tokens = f.key.split(' ')
    let hits = 0
    for (const t of tokens) if (wantedTokens.has(t)) hits++
    if (hits === 0) continue
    const score = hits / wantedTokens.size - (tokens.length - hits) * 0.08
    scored.push({ f, score })
  }
  scored.sort((a, b) => b.score - a.score || a.f.name.length - b.f.name.length)
  return scored.slice(0, limit).map(({ f }) => f)
}

/* ------------------------------------------------------------------ */
/* Search                                                              */
/* ------------------------------------------------------------------ */

export type SortKey = 'popular' | 'name' | 'newest'

export interface Query {
  /** Free text; every whitespace-separated token must appear in the name. */
  text: string
  categories: Category[]
  sources: Source[]
  /** SPDX ids; empty = any. */
  licenses: string[]
  /** `latin`, `cyrillic`, ...; empty = any. Every listed subset must be covered. */
  subsets: string[]
  variableOnly: boolean
  /** Only families that have an italic. */
  italicOnly: boolean
  /** Only families with a file to download (a few Google entries have none). */
  downloadableOnly: boolean
  sort: SortKey
}

export const EMPTY_QUERY: Query = {
  text: '',
  categories: [],
  sources: [],
  licenses: [],
  subsets: [],
  variableOnly: false,
  italicOnly: false,
  downloadableOnly: true,
  sort: 'popular',
}

export function searchFamilies(q: Query, pool: readonly Family[] = FAMILIES): Family[] {
  const tokens = normalizeKey(q.text).split(' ').filter(Boolean)
  const out: Family[] = []
  for (const f of pool) {
    if (q.downloadableOnly && f.files.length === 0) continue
    if (q.categories.length && !q.categories.includes(f.category)) continue
    if (q.sources.length && !q.sources.includes(f.source)) continue
    if (q.licenses.length && !q.licenses.includes(f.license)) continue
    if (q.variableOnly && f.axes.length === 0) continue
    if (q.italicOnly && !f.italic) continue
    if (q.subsets.length && !q.subsets.every((s) => f.subsets.includes(s))) continue
    if (tokens.length) {
      const hay = `${f.key} ${f.designers.join(' ').toLowerCase()}`
      if (!tokens.every((t) => hay.includes(t))) continue
    }
    out.push(f)
  }
  sortFamilies(out, q.sort, tokens)
  return out
}

function sortFamilies(list: Family[], sort: SortKey, tokens: string[]): void {
  if (sort === 'name') {
    list.sort((a, b) => a.name.localeCompare(b.name))
    return
  }
  if (sort === 'newest') {
    list.sort((a, b) => b.added.localeCompare(a.added) || a.name.localeCompare(b.name))
    return
  }
  // Popular — but a name that *starts* with the search text outranks one that
  // merely contains it, so typing "Roboto" puts Roboto above Roboto Slab and
  // both above anything whose designer happens to mention it.
  const prefix = tokens.join(' ')
  const starts = (f: Family) => (prefix && f.key.startsWith(prefix) ? 0 : 1)
  list.sort((a, b) => starts(a) - starts(b) || a.rank - b.rank || a.name.localeCompare(b.name))
}

/** Every distinct licence in the catalogue, most common first. */
export function licenseOptions(): Array<{ id: string; count: number }> {
  const counts = new Map<string, number>()
  for (const f of FAMILIES) counts.set(f.license, (counts.get(f.license) ?? 0) + 1)
  return [...counts].map(([id, count]) => ({ id, count })).sort((a, b) => b.count - a.count)
}

/** Every distinct subset, most common first. */
export function subsetOptions(): Array<{ id: string; count: number }> {
  const counts = new Map<string, number>()
  for (const f of FAMILIES) for (const s of f.subsets) counts.set(s, (counts.get(s) ?? 0) + 1)
  return [...counts].map(([id, count]) => ({ id, count })).sort((a, b) => b.count - a.count)
}

File: src-tauri/src/catalogue.rs
//! The font catalogue, embedded at build time so the headless sync pass can
//! resolve a family name to files without a webview.
//!
//! Same two JSON snapshots the frontend ships (`src/data/*.json`), read via
//! `include_str!` so the two sides can never disagree about what exists. The
//! matching rule mirrors `findFamily` in `src/core/catalogue.ts`: normalised
//! name first, then the name with a trailing style word stripped.

use std::collections::HashMap;
use std::sync::OnceLock;

use serde::Deserialize;

const GOOGLE_JSON: &str = include_str!("../../src/data/google-fonts.json");
const FONTSOURCE_JSON: &str = include_str!("../../src/data/fontsource-fonts.json");

const RAW_BASE: &str = "https://raw.githubusercontent.com/google/fonts/main";
const GH_MIRROR: &str = "https://cdn.jsdelivr.net/gh/google/fonts@main";
const FONTSOURCE_CDN: &str = "https://cdn.jsdelivr.net/fontsource/fonts";

#[derive(Deserialize)]
struct RawGoogle {
    n: String,
    #[serde(default)]
    l: String,
    #[serde(default)]
    p: String,
    #[serde(default)]
    f: Vec<String>,
    #[serde(default)]
    w: Vec<u32>,
}

#[derive(Deserialize)]
struct RawFontsource {
    n: String,
    s: String,
    u: String,
    #[serde(default)]
    w: Vec<u32>,
    #[serde(default)]
    i: u8,
    #[serde(default)]
    l: String,
}

#[derive(Deserialize)]
struct Wrapper<T> {
    families: Vec<T>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Source {
    Google,
    Fontsource,
}

#[derive(Clone, Debug)]
pub struct File {
    pub filename: String,
    pub url: String,
    pub mirrors: Vec<String>,
    pub weight: u32,
    pub italic: bool,
    pub variable: bool,
}

#[derive(Clone, Debug)]
pub struct Family {
    pub id: String,
    pub name: String,
    pub source: Source,
    pub license: String,
    pub license_url: Option<String>,
    pub weights: Vec<u32>,
    pub files: Vec<File>,
    key: String,
}

/// Case/punctuation-insensitive key — the same rule as `normalizeKey` in
/// `src/core/names.ts`.
pub fn normalize_key(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut last_space = true;
    for ch in s.chars() {
        if matches!(ch, '\u{2018}' | '\u{2019}' | '\u{201C}' | '\u{201D}') {
            continue;
        }
        let c = if ch == '-' || ch == '_' || ch == ',' {
            ' '
        } else {
            ch
        };
        if c.is_whitespace() {
            if !last_space {
                out.push(' ');
                last_space = true;
            }
        } else {
            for lower in c.to_lowercase() {
                out.push(lower);
            }
            last_space = false;
        }
    }
    out.trim().to_string()
}

/// Compound forms first, so `semi bold` is taken whole rather than as `bold`.
const STYLE_WORDS: &[&str] = &[
    "extra light",
    "ultra light",
    "semi bold",
    "demi bold",
    "extra bold",
    "ultra bold",
    "thin",
    "hairline",
    "extralight",
    "ultralight",
    "light",
    "regular",
    "normal",
    "book",
    "roman",
    "medium",
    "semibold",
    "demibold",
    "bold",
    "extrabold",
    "ultrabold",
    "heavy",
    "black",
    "italic",
    "oblique",
];

/// Strip trailing style words: `poppins semi bold italic` -> `poppins`.
///
/// Conservative, like the frontend's `parseFontName`: a name that is *only*
/// style words is kept whole, so the families called `Black` and `Medium`
/// still resolve, and a word boundary is required so `bookman` keeps its
/// `book`.
pub fn strip_style(key: &str) -> String {
    let mut work = key.to_string();
    loop {
        let mut stripped = false;
        for w in STYLE_WORDS {
            if let Some(head) = work.strip_suffix(w) {
                if !head.ends_with(' ') {
                    continue;
                }
                work = head.trim_end().to_string();
                stripped = true;
                break;
            }
        }
        if !stripped {
            break;
        }
    }
    // Numeric weights: `roboto 700`.
    if let Some(idx) = work.rfind(' ') {
        let tail = &work[idx + 1..];
        if tail.len() == 3 && tail.ends_with("00") && tail.as_bytes()[0].is_ascii_digit() {
            work = work[..idx].trim_end().to_string();
        }
    }
    if work.is_empty() {
        key.to_string()
    } else {
        work
    }
}

struct Catalogue {
    families: Vec<Family>,
    by_key: HashMap<String, usize>,
    by_id: HashMap<String, usize>,
}

fn describe_google_file(file: &str) -> (u32, bool, bool) {
    let base = file.strip_suffix(".ttf").unwrap_or(file);
    let variable = base.ends_with(']') && base.contains('[');
    let stem = if variable {
        &base[..base.rfind('[').unwrap_or(base.len())]
    } else {
        base
    };
    let suffix = stem.find('-').map(|i| &stem[i + 1..]).unwrap_or("");
    let lower = suffix.to_ascii_lowercase();
    let italic = lower.contains("italic");
    let without = lower.replace("italic", "");
    let weight = match without.as_str() {
        "thin" | "hairline" => 100,
        "extralight" | "ultralight" => 200,
        "light" => 300,
        "medium" => 500,
        "semibold" | "demibold" => 600,
        "bold" => 700,
        "extrabold" | "ultrabold" => 800,
        "black" | "heavy" => 900,
        _ => 400,
    };
    (weight, italic, variable)
}

fn slug(name: &str) -> String {
    let mut out = String::new();
    let mut dash = false;
    for ch in name.to_ascii_lowercase().chars() {
        if ch.is_ascii_alphanumeric() {
            out.push(ch);
            dash = false;
        } else if !dash && !out.is_empty() {
            out.push('-');
            dash = true;
        }
    }
    out.trim_end_matches('-').to_string()
}

fn encode(s: &str) -> String {
    // Only the characters that actually appear in google/fonts filenames need
    // escaping: `[`, `]` and `,` in variable-font names.
    let mut out = String::with_capacity(s.len());
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(b as char)
            }
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}

fn weight_name(w: u32) -> String {
    match w {
        100 => "Thin",
        200 => "ExtraLight",
        300 => "Light",
        400 => "Regular",
        500 => "Medium",
        600 => "SemiBold",
        700 => "Bold",
        800 => "ExtraBold",
        900 => "Black",
        _ => return w.to_string(),
    }
    .to_string()
}

fn load() -> Catalogue {
    let google: Wrapper<RawGoogle> =
        serde_json::from_str(GOOGLE_JSON).expect("google-fonts.json is valid");
    let fontsource: Wrapper<RawFontsource> =
        serde_json::from_str(FONTSOURCE_JSON).expect("fontsource-fonts.json is valid");

    let mut families = Vec::with_capacity(google.families.len() + fontsource.families.len());
    for f in google.families {
        let license_url = if f.p.is_empty() {
            None
        } else {
            let file = match f.p.split('/').next().unwrap_or("ofl") {
                "apache" => "LICENSE.txt",
                "ufl" => "UFL.txt",
                _ => "OFL.txt",
            };
            Some(format!("{RAW_BASE}/{}/{file}", f.p))
        };
        let files =
            f.f.iter()
                .map(|file| {
                    let (weight, italic, variable) = describe_google_file(file);
                    let enc = encode(file);
                    File {
                        filename: file.clone(),
                        url: format!("{RAW_BASE}/{}/{enc}", f.p),
                        mirrors: vec![format!("{GH_MIRROR}/{}/{enc}", f.p)],
                        weight,
                        italic,
                        variable,
                    }
                })
                .collect();
        families.push(Family {
            id: format!("google:{}", slug(&f.n)),
            key: normalize_key(&f.n),
            name: f.n,
            source: Source::Google,
            license: if f.l.is_empty() {
                "Unknown".into()
            } else {
                f.l
            },
            license_url,
            weights: f.w,
            files,
        });
    }
    for f in fontsource.families {
        let stem: String = f.n.chars().filter(|c| c.is_ascii_alphanumeric()).collect();
        let weights = if f.w.is_empty() {
            vec![400]
        } else {
            f.w.clone()
        };
        let mut files = Vec::new();
        for &w in &weights {
            for italic in [false, true] {
                if italic && f.i == 0 {
                    continue;
                }
                let wn = weight_name(w);
                let style = if italic {
                    if w == 400 {
                        "Italic".to_string()
                    } else {
                        format!("{wn}Italic")
                    }
                } else {
                    wn
                };
                files.push(File {
                    filename: format!("{stem}-{style}.ttf"),
                    url: format!(
                        "{FONTSOURCE_CDN}/{}@latest/{}-{w}-{}.ttf",
                        f.s,
                        f.u,
                        if italic { "italic" } else { "normal" }
                    ),
                    mirrors: Vec::new(),
                    weight: w,
                    italic,
                    variable: false,
                });
            }
        }
        families.push(Family {
            id: format!("fontsource:{}", f.s),
            key: normalize_key(&f.n),
            name: f.n,
            source: Source::Fontsource,
            license: f.l,
            license_url: Some(format!(
                "https://cdn.jsdelivr.net/npm/@fontsource/{}@latest/LICENSE",
                f.s
            )),
            weights,
            files,
        });
    }

    let mut by_key = HashMap::new();
    let mut by_id = HashMap::new();
    for (i, f) in families.iter().enumerate() {
        by_key.entry(f.key.clone()).or_insert(i);
        by_id.insert(f.id.clone(), i);
    }
    Catalogue {
        families,
        by_key,
        by_id,
    }
}

fn catalogue() -> &'static Catalogue {
    static CAT: OnceLock<Catalogue> = OnceLock::new();
    CAT.get_or_init(load)
}

pub fn count() -> usize {
    catalogue().families.len()
}

pub fn by_id(id: &str) -> Option<&'static Family> {
    let c = catalogue();
    c.by_id.get(id).map(|&i| &c.families[i])
}

/// Loose lookup: `Poppins`, `poppins-bold`, `Poppins SemiBold Italic`.
pub fn find(name: &str, source: Option<Source>) -> Option<&'static Family> {
    let c = catalogue();
    let key = normalize_key(name);
    let stripped = strip_style(&key);
    for k in [key.as_str(), stripped.as_str()] {
        if let Some(&i) = c.by_key.get(k) {
            let f = &c.families[i];
            if source.as_ref().map_or(true, |s| *s == f.source) {
                return Some(f);
            }
        }
    }
    if let Some(s) = source {
        for k in [key.as_str(), stripped.as_str()] {
            if let Some(f) = c.families.iter().find(|f| f.source == s && f.key == k) {
                return Some(f);
            }
        }
    }
    None
}

/// Files for a requested selection. `weights == None` means the whole family.
pub fn select_files(family: &Family, weights: Option<&[u32]>, italics: bool) -> Vec<File> {
    let Some(weights) = weights else {
        return family.files.clone();
    };
    let mut out: Vec<File> = Vec::new();
    let styles: &[bool] = if italics { &[false, true] } else { &[false] };
    let weights: Vec<u32> = if weights.is_empty() {
        vec![400]
    } else {
        weights.to_vec()
    };
    for &w in &weights {
        for &italic in styles {
            let pick = pick_file(&family.files, w, italic);
            if let Some(p) = pick {
                if !out.iter().any(|o| o.filename == p.filename) {
                    out.push(p.clone());
                }
            }
        }
    }
    out
}

fn pick_file(files: &[File], want: u32, italic: bool) -> Option<&File> {
    let mut statics: Vec<&File> = files
        .iter()
        .filter(|f| !f.variable && f.italic == italic)
        .collect();
    if !statics.is_empty() {
        statics.sort_by_key(|f| (f.weight as i64 - want as i64).abs());
        return statics.first().copied();
    }
    if let Some(v) = files.iter().find(|f| f.variable && f.italic == italic) {
        return Some(v);
    }
    files.iter().find(|f| !f.italic).or_else(|| files.first())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn loads_both_sources() {
        assert!(count() > 2000);
        assert!(by_id("google:poppins").is_some());
        assert!(by_id("fontsource:adwaita-sans").is_some());
    }

    #[test]
    fn strips_styles_conservatively() {
        assert_eq!(strip_style("poppins semi bold italic"), "poppins");
        assert_eq!(strip_style("poppins bold"), "poppins");
        assert_eq!(strip_style("roboto 700"), "roboto");
        assert_eq!(strip_style("black"), "black");
        assert_eq!(strip_style("bookman old style"), "bookman old style");
        assert_eq!(strip_style("archivo black"), "archivo");
    }

    #[test]
    fn finds_loosely() {
        assert_eq!(find("poppins", None).unwrap().id, "google:poppins");
        assert_eq!(find("Poppins-Bold", None).unwrap().id, "google:poppins");
        assert_eq!(
            find("Poppins SemiBold Italic", None).unwrap().id,
            "google:poppins"
        );
        assert_eq!(
            find("adwaita sans", None).unwrap().source,
            Source::Fontsource
        );
        assert!(find("Poppins", Some(Source::Fontsource)).is_none());
        assert!(find("Calibri", None).is_none());
        // `Archivo Black` is a real family; `Black` alone is too.
        assert_eq!(find("Archivo Black", None).unwrap().name, "Archivo Black");
    }

    #[test]
    fn builds_urls_like_the_frontend() {
        let p = find("Poppins", None).unwrap();
        let semi = p
            .files
            .iter()
            .find(|f| f.filename == "Poppins-SemiBold.ttf")
            .unwrap();
        assert_eq!(
            semi.url,
            "https://raw.githubusercontent.com/google/fonts/main/ofl/poppins/Poppins-SemiBold.ttf"
        );
        assert_eq!(
            semi.mirrors[0],
            "https://cdn.jsdelivr.net/gh/google/fonts@main/ofl/poppins/Poppins-SemiBold.ttf"
        );
        assert_eq!(
            (semi.weight, semi.italic, semi.variable),
            (600, false, false)
        );
        assert_eq!(
            p.license_url.as_deref(),
            Some("https://raw.githubusercontent.com/google/fonts/main/ofl/poppins/OFL.txt")
        );

        let inter = find("Inter", None).unwrap();
        let var = inter
            .files
            .iter()
            .find(|f| f.filename == "Inter[opsz,wght].ttf")
            .unwrap();
        assert!(var.variable && !var.italic);
        assert!(var.url.ends_with("/ofl/inter/Inter%5Bopsz%2Cwght%5D.ttf"));

        let a = find("Adwaita Sans", None).unwrap();
        assert!(a
            .files
            .iter()
            .any(|f| f.filename == "AdwaitaSans-Regular.ttf"));
        assert!(a.files[0]
            .url
            .starts_with("https://cdn.jsdelivr.net/fontsource/fonts/adwaita-sans@latest/latin-"));
    }

    #[test]
    fn selects_files() {
        let p = find("Poppins", None).unwrap();
        let all = select_files(p, None, true);
        assert_eq!(all.len(), p.files.len());
        let some = select_files(p, Some(&[400, 700]), true);
        let names: Vec<&str> = some.iter().map(|f| f.filename.as_str()).collect();
        assert_eq!(
            names,
            [
                "Poppins-Regular.ttf",
                "Poppins-Italic.ttf",
                "Poppins-Bold.ttf",
                "Poppins-BoldItalic.ttf"
            ]
        );
        let inter = find("Inter", None).unwrap();
        let v = select_files(inter, Some(&[300]), false);
        assert_eq!(v.len(), 1);
        assert!(v[0].variable);
    }
}
